### PR TITLE
Bugfix: Validation papercuts

### PR DESCRIPTION
**Pull request by eureton**
Notes:
- `(s/id)`: `pack-doc` doesn't leave malformed values out
- `(s/map)`: fixes validation of non-map values
    ``` clj
    => (def dtc
         (c/dom-type-collection :name :xyz
                                :properties {:v (s/map {:id (s/id)})}))
    => (v/validate-properties false (:properties dtc) {:v 20 :company-id (ObjectId.)})
    [false {:v "not sequential"}]
    ```
- `(s/vector)`:
  - non-vector values:
    - fixes validation
      ``` clj
      => (def dtc
           (c/dom-type-collection :name :xyz
                                  :properties {:v (s/vector (s/number))}))
      => => (v/validate-properties false (:properties dtc) {:v 20 :company-id (ObjectId.)})
      [false {:v "not a map", :id "required property has no value"}]
      ```
  - vector values:
    - validates each item separately
    - collects item error messages in a hash, keyed by index
      ``` clj
      => (def dtc
           (c/dom-type-collection :name :xyz
                                  :properties {:v (s/vector (s/number))}))
      => (v/validate-properties false (:properties dtc) {:v [10 "20"] :company-id (ObjectId.)})
      [false {:v {1 "not a number"}}]
      ```
- `:min-length` / `:max-length`: apply to `:string`, `:vector`, `:map` only

### CODE REVIEW

File: domain/test/com/timezynk/domain/schema/map_test.clj [hunk 1]
```diff
@@ -0,0 +1,65 @@
+(ns com.timezynk.domain.schema.map-test
+  (:require [clojure.test :refer [deftest is testing use-fixtures]]
+            [com.timezynk.assembly-line :as a]
+            [com.timezynk.domain.core :as c]
+            [com.timezynk.domain.persistence :as p]
+            [com.timezynk.domain.schema :as s])
+  (:use [slingshot.test])
```
Comment (LarsMelander) [inline domain/test/com/timezynk/domain/schema/map_test.clj:7]: Don't use `:use`, use `:require`.

File: domain/test/com/timezynk/domain/schema/map_test.clj [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+(deftest valid
+  (is (insert! valid-doc)))
+
+(deftest map-value-not-matching-key-type
+  (testing "type mismatch within map"
+    (let [invalid-doc-1 (update-in valid-doc [:x :id] str)
+          invalid-doc-2 (update-in valid-doc [:x :ref-no] str)
+          invalid-doc-3 (assoc-in valid-doc [:x :sold] 1)]
+      (testing "string instead of ObjectId"
+        (is (thrown+? (= (get-in % [:errors :id]) "not a valid id")
```
Comment (LarsMelander) [inline domain/test/com/timezynk/domain/schema/map_test.clj:44]: All lines with `thrown?` return errors.

**Comment by jheander** [comment]
12 tests fail when running `lein test`, but all test pass in `lein kaocha`. Using `Leiningen 2.9.5 on Java 11.0.9.1 OpenJDK 64-Bit Server VM`

**Comment by eureton** [comment]
@jheander Indeed, the way tests are setup exposes a `leiningen` bug in versions before `2.9.8`.

The error manifests as a failure to initialize DTCs properly. Despite being freshly returned by `com.timezynk.domain.core/dom-type-collection`, the DTC is allegedly **not** supporting the `com.timezynk.domain.persistence/Persistence` protocol. Looking at the implementation of `dom-type-collection`, it is clearly impossible to get something not supporting `Persistence` 😕.

Running the very same test with `lein` versions `2.9.8` and `2.9.10` works fine.

Last but not least, `leiningen` versions older than `2.9.8` are not breaking code in this PR only. I have checked the rest of the similarly created DTC variables (we have 3 of them) and they demonstrate the same problem. The reason we haven't noticed is that they happen to not be used through the `Persistence` protocol.

**Comment by jheander** [comment]
> @jheander Indeed, the way tests are setup exposes a `leiningen` bug in versions before `2.9.8`.
> 
> The error manifests as a failure to initialize DTCs properly. Despite being freshly returned by `com.timezynk.domain.core/dom-type-collection`, the DTC is allegedly **not** supporting the `com.timezynk.domain.persistence/Persistence` protocol. Looking at the implementation of `dom-type-collection`, it is clearly impossible to get something not supporting `Persistence` 😕.
> 
> Running the very same test with `lein` versions `2.9.8` and `2.9.10` works fine.
> 
> Last but not least, `leiningen` versions older than `2.9.8` are not breaking code in this PR only. I have checked the rest of the similarly created DTC variables (we have 3 of them) and they demonstrate the same problem. The reason we haven't noticed is that they happen to not be used through the `Persistence` protocol.

Great find! I have pushed upgrades to leiningen 2.9.10 to the devcontainer and CI build environments.

**Pull request merged**